### ISSUE TITLE
Recent changes in PMIx tool I/O and event handling causing PRRTE debug example failures?

### ISSUE THREAD

**Issue by drwootton** [openpmix/prrte#1019]
Thank you for taking the time to submit an issue!

### What version of the PMIx Reference Server are you using? (e.g., v1.0, v2.1, git master @ hash, etc.)
Current PRRTE master source

##### What version of PMIx are you using? (e.g., v1.2.5, v2.0.3, v2.1.0, git branch name and hash, etc.)
Current PMIx master source

### Please describe the system on which you are running

* Operating system/version:
RHEL 7.6

* Computer hardware:
Power 9, single node, 3 nodes

* Network type:
Ethernet

-----------------------------

## Details of the problem

The PRRTE examples/debugger indirect and indirect-multi exapmples are not displaying stdout output from the application 'hello' process after recent PMIx and PRRTE changes. Also both the handler I registered for the PMIX_ERR_LOST_CONNECTION event and the default handler were getting invoked for this event. Now only the handler I registered is getting called.

I run the indirect example as **indirect prterun -n 2 ./hello 10**

I run the indirect-multi example as **indirect-multi --daemon-colocate-per-node 1 prterun --hostfile hostfile --np 12 ./hello 10** where hostfile lists 3 nodes with 4 slots each.

In both cases output for the application hello processes is no longer displayed.

These examples do not issue any PMIx_IOF* calls. Even without these calls, these examples used to display stdout output from the hello processes. I assumed, maybe incorrectly, that the PMIx_IOF* calls were not required, and that stdio output from the application processes was somehow being passed thru prterun and back to the tool front end process.

I noticed a few pull requests related to I/O forwarding and am wondering if something changed in how application process stdio was handled and if I really am required to use PMIx I/O forwarding for this to work right.

These PMIX_ERR_LOST_CONNECTION event handling for these testcases, as well as some other examples, changed such that only the handler I registered to handle the event runs and the default handler no longer runs. My lost connection handler calls the cbfunc callback chain handler passing PMIX_EVENT_ACTION_COMPLETE.

It looks like PMIx issue 2220 fixed a problem with this handling and I should not be expecting the default handler to run in that cast.

**Comment by rhc54** [comment]
Please update both PMIx and PRRTE. I have updated the `indirect-multi` example so it works correctly - just a couple of minor errors in it. Please use it as a template for updating the other examples and I believe they should be working.

**Comment by jjhursey** [comment]
Ref #1026

**Comment by drwootton** [comment]
I got the latest indirect-multi.c source from github as well as updated PMIx and PRRTE. When I run any of the indirect-multi tests I get these messages written to stderr and the test terminates.
~~~
[f7n01:14591] PMIX b:unpack: got type 40 when expecting type 5
[f7n01:14591] PMIX b:unpack: got type 40 when expecting type 5
[f7n01:14591] PMIX b:unpack: got type 40 when expecting type 5
[f7n01:14591] [base/base_receive.c:467] PMIx Error: PACK-MISMATCH
[f7n01:14591] [base/base_receive.c:467] PMIx Error: PACK-MISMATCH
[f7n01:14591] [base/base_receive.c:467] PMIx Error: PACK-MISMATCH
indirect-multi      : Error: Launcher failed
~~~
None of the other examples, which I have not started updating, have this problem.

**Comment by rhc54** [comment]
Looks to me like you have something borked in your setup - that looks like what I'd expect to see if something was using a different version of the PMIx library. The indirect-multi.c source should be taken from PRRTE's example/debugger directory.

**Comment by drwootton** [comment]
That's what I suspected at first too,so I've spent the afternoon trying to track this down. I have made sure that the only thing I had in my PATH and LD_LIBRARY_PATH related to PMIx, PRRTE or OpenMPI is the PMIx and PRRTE which I built from source where I only run git pulls and no changes of my own. Also, running ldd on hello, prterun and the daemon, both on the local node and running ssh to another node to make sure I wasn't picking up something in a profile script only showed my PMIx and PRRTE directories. I also pulled indirect-multi source from github

**Comment by drwootton** [comment]
I first suspected something wrong with my environment so I spent the afternoon trying to track this down. I checked all my login/profile scripts and made sure that only PATH was set and that it was set to only the PMIX and PRRTE directories I built from source where I only run git pull and never edit. I am also using the current indirect-multi.c from github.

I ran ldd both on the local node and via ssh for prterun and prted and both are only pointing to my directories.

Can you suggest a debug option I can set to try to track this down?

**Comment by rhc54** [comment]
Not sure how to advise - somehow, you are picking up something stale. Did you make sure to `rm -rf` your prior installations before building the new PMIx and PRRTE clones? Don't just do a `git pull` in existing clones when you see this happen - start completely from scratch to ensure things are clean.

**Comment by jjhursey** [comment]
I found the issue and posted a fix in PR #1028

**Comment by drwootton** [comment]
PR #1028 does fix the error messages. However, with the current indirect-multi.c from the Github master branch, unmodified, I am still not seeing the output from the hello target application in the output that comes back to my terminal session. I do see the debugger daemon output.

I did see output from the hello target in my terminal session for this and the other examples except attach previously. I'm not sure if the hello stdio output is supposed to appear and something is broken, or if this is the way the code is supposed to work.

**Comment by drwootton** [comment]
It looks like this was my problem. @jjhursey found my PRRTE code was backlevel. That did not solve all my problems with hello's stdio, so found I was running with ksh as my login shell, which I have been using forever. I changed it to use bash and now indirect-multi is working, including showing me hello's stdout. I have no idea what that has to do with anything, but it works, so closing this issue again.

**Comment by rhc54** [comment]
> I was running with ksh as my login shell,

If you just run `prterun -n 1  hostname` or some other appropriate app under ksh, do you see the output?

**Comment by drwootton** [comment]
I was getting normal output for any command like prterun. The only exception was that when I ran the debugger examples, the hello process output was lost, where the tool, the daemon, and the hello process were all writing to stdout. 

I was working with @jjhursey to figure out my problems, and was still having problems after resolving my backlevel code, but he was not. I noticed his shell was bash and mine was ksh, so I decided to try changing my shell and I started getting hello process output.

After that, I repeatedly ran all of the examples successfully. 

I changed my shell on a single node back to ksh and the example still worked, so maybe I inadvertently fixed something else.

Losing output consistently from one application with ksh vs bash makes absolutely no sense, but something changed.

Since everything is working and I've never seen this behavior before with ksh I consider this a fluke that is resolved.
